include/cppkafka/callback_invoker.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2017, Matias Fontanini
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * * Redistributions in binary form must reproduce the above
+ *   copyright notice, this list of conditions and the following disclaimer
+ *   in the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef CPPKAFKA_CALLBACK_INVOKER_H
+#define CPPKAFKA_CALLBACK_INVOKER_H
+
+#include <sstream>
+#include <assert.h>
+#include "logging.h"
+#include "kafka_handle_base.h"
+
+namespace cppkafka {
+
+// Success values
+template <typename T>
+T success_value() { return T(); }
+
+template<> inline
+void success_value<void>() {};
+
+template<> inline
+bool success_value<bool>() { return true; }
+
+// Error values
+template <typename T>
+T error_value() { return T(); }
+
+template<> inline
+void error_value<void>() {};
+
+template<> inline
+bool error_value<bool>() { return false; }
+
+template<> inline
+int error_value<int>() { return -1; }
+
+// CallbackInvoker
+template <typename RetType, typename ...Args>
+class CallbackInvoker;
+
+template <typename RetType, typename ...Args>
+class CallbackInvoker<RetType(Args...)>
```

Comment:
Wouldn't it be better to use a simple function instead of this? I don't think there should be a need to indicate all template parameters when instantiating this, especially since this is always used internally. Something like this would work:

```c++
template <typename RetType, typename Callback, typename... Args>
RetType safe_invoke_callback(const std::string& callback_name,
                                                  const Callback& callback,
                                                  KafkaHandleBase* handle,
                                                  Args&&... args);
``` 
Which you can just invoke by doing:

```c++
safe_invoke_callback<int32_t>(callback_name, callback, nullptr,
                                                   topic, key, partition_count);
```